feat: stickied messages

DDbrother9999 opened the pull request:
This PR adds the ability for staff members to sticky messages in order for them to always be visible in channels. It's intended to be used during downtime when the server is not in lockdown, but when a continual message being prompted is helpful. 

[see discord discussion](https://discord.com/channels/1142987286074642474/1455277581824954378)

<!-- This is an auto-generated description by cubic. -->
---
## Summary by cubic
Adds stickied messages so staff can keep an info message visible in a channel during downtime. Messages are sent as embeds (title for short text, body for long) with a footer, and managed via the /sticky command.

- **New Features**
  - /sticky subcommands: start (optional message), stop, set (required message), view
  - Requires Manage Messages; guild-only; replies are ephemeral
  - One sticky per channel; supports multiple channels concurrently
  - Background job checks every 5s, deletes the old sticky, and re-posts to keep it at the bottom

<sup>Written for commit eecf09f3995abba89dd60f41ac5e206642a58511. Summary will update on new commits.</sup>

<!-- End of auto-generated description by cubic. -->

DDbrother9999 committed:
feat: sticky messages

DDbrother9999 committed:
fix: small optimization for stickied messages

DDbrother9999 committed:
fix: no guildId

idk what I was thinking, also fixed multi-channel sticks

DDbrother9999 committed:
chore: kdoc StickyManager

DDbrother9999 committed:
fix: safely handle exceptions within sticky loop

DDbrother9999 committed:
fix: context subtext on sticky messages

DDbrother9999 committed:
fix: transition to an embed for the stickied message

the message is purposely left blank, instead, the content is placed in the title field

tarna commented on src/main/kotlin/me/santio/minehututils/sticky/StickyManager.kt:
```diff
@@ -0,0 +1,93 @@
+
+    private fun startLoop() {
+        scope.launch {
+            while (true) {
```
Rather than doing a while loop like this, you should use timer.schedule similar to how its done in [MinehutUtils.kt](https://github.com/SantioMC/MinehutUtils/blob/master/src/main/kotlin/me/santio/minehututils/MinehutUtils.kt#L86-L97) from lines 86 to 97 to refresh other data.

tarna commented on src/main/kotlin/me/santio/minehututils/commands/impl/StickyCommand.kt:
```diff
@@ -0,0 +1,78 @@
+class StickyCommand : SlashCommand {
+
+    override fun getData(): CommandData {
+        return Command("sticky", "Sticky a message") {
```
I think a better command for this would be `/sticky start [message]`, `/sticky set <message>`, and `/sticky stop`. Start would start a sticky message in the current channel with a message. Set will allow you to change the message. And stop would just stop it from being active, but it should still store the message, so if you were to do `/stick start` again, it would start it up without having to retype the message again. And maybe even a view command to view the current sticky message while its not active.

PrecedingxD replied:
100% agree with this

tarna commented on src/main/kotlin/me/santio/minehututils/commands/impl/StickyCommand.kt:
```diff
@@ -0,0 +1,78 @@
+            }
+
+            if (StickyManager.isStickied(channel.id)) {
+                event.replyEmbeds(
```
You should also log a message to the logs channel whenever a sticky message is started or stopped. You can view some of the other commands that use GuildLogger to see how it is done.

DDbrother9999 committed:
fix: split into subcommands (start, stop, set, view)

DDbrother9999 committed:
fix: use timer.schedule

DDbrother9999 committed:
fix: proper logging (start, stop, set)

DDbrother9999 committed:
remove userid from sticky, lint in StickyManager

DDbrother9999 committed:
fix: userid from StickyCommand

DDbrother9999 committed:
chore: cleanup
also fix edgecase when length < 256 and < 4096

tarna commented on src/main/kotlin/me/santio/minehututils/commands/impl/StickyCommand.kt:
```diff
@@ -0,0 +1,122 @@
+                        Option<String>("message", "The message to stick", required = false)
+                    )
+                },
+                Subcommand("stop", "Unstick the message") {},
```
You don't need to add the empty brackets. It'll work fine without it.

tarna commented on src/main/kotlin/me/santio/minehututils/commands/impl/StickyCommand.kt:
```diff
@@ -0,0 +1,122 @@
+                        Option<String>("message", "The message to stick", required = true)
+                    )
+                },
+                Subcommand("view", "View the stickied message") {
```
Same for the brackets here.

tarna commented on src/main/kotlin/me/santio/minehututils/sticky/StickyManager.kt:
```diff
@@ -0,0 +1,135 @@
+     * @param channelId The id of the channel to sticky
+     * @param message The message to sticky
+     */
+    fun start(channelId: String, message: String?) {
```
Is there a reason the message should ever be null? You have a check in the start command to display an error already if the message is null. So this should always be set.

tarna commented on src/main/kotlin/me/santio/minehututils/commands/impl/StickyCommand.kt:
```diff
@@ -0,0 +1,122 @@
+            }
+
+            "stop" -> {
+                if (!StickyManager.isActive(channel.id)) error("There is not a stickied message in this channel.", true)
```
The second parameter of error is already true by default, so you can just remove that. There are a couple other places you do this as well.

tarna commented on src/main/kotlin/me/santio/minehututils/commands/impl/StickyCommand.kt:
```diff
@@ -0,0 +1,122 @@
+                        guild
+                    ).build()
+                ).setEphemeral(true).queue()
+                GuildLogger.of(event.guild!!).log(
```
You have a guild variable you are already using above. You should use that again here.

tarna commented on src/main/kotlin/me/santio/minehututils/commands/impl/StickyCommand.kt:
```diff
@@ -0,0 +1,122 @@
+            }
+
+            "set" -> {
+                val message = event.getOption("message")?.asString
```
Even though the message should always be set, you should still do ?: error as thats what other commands do as well with required options. And you won't have to do message!!.

tarna commented on src/main/kotlin/me/santio/minehututils/sticky/StickyManager.kt:
```diff
@@ -0,0 +1,135 @@
+        val channelId: String,
+        var message: String,
+        var lastMessageId: String? = null,
+        var active: Boolean? = false
```
Active shouldn't be nullable here as it is by default false.

DDbrother9999 committed:
fix: readability + null safety things

DDbrother9999 committed:
chore: lint

tarna commented on src/main/kotlin/me/santio/minehututils/sticky/StickyManager.kt:
```diff
@@ -0,0 +1,136 @@
+                active = true
+            )
+        } else {
+            message.let { sticky.message = it }
```
You shouldn't need to do message.let here anymore as message is always set.

tarna commented on src/main/kotlin/me/santio/minehututils/sticky/StickyManager.kt:
```diff
@@ -0,0 +1,136 @@
+    data class StickyMessage(
+        val channelId: String,
+        var message: String,
+        var lastMessageId: String? = null,
```
May also be worth it to make lastMessageId not nullable by sending the embed immediately when a sticky is started and then storing that so it is always set. Make sure to also remove any .let you have done on this value if you make this change.

DDbrother9999 replied:
This would break the /sticky set without starting the sticky flow. I'm not sure that's worth preserving though. In leiu of this, I've added a forceRefresh method to fix the behavior of starting/setting without immediately updating the message (this would also occur in the case that a message was updated, but the last message in the channel == lastMessageId preventing an update).

This also renders the double embeds after a /set obselete so, that's also been removed. imo that's cleaner

In suggesting this though, I fixed an issue where deleting the last embed would just brick the whole loop since it would fail to delete the last message itself and just never send a new one :/

tarna commented on src/main/kotlin/me/santio/minehututils/commands/impl/StickyCommand.kt:
```diff
@@ -0,0 +1,123 @@
+
+    override suspend fun execute(event: SlashCommandInteractionEvent) {
+        val channel = event.channel
+        val guild = event.guild
```
Just realized that the other commands just do ?: return if guild isn't set, so you should probably do that here as well. So you won't have to do !! each time you want to use it.

tarna commented on src/main/kotlin/me/santio/minehututils/commands/impl/StickyCommand.kt:
```diff
@@ -0,0 +1,123 @@
+        val channel = event.channel
+        val guild = event.guild
+
+        when (event.subcommandName) {
```
Similar to the lockdown and cooldown commands, you should have a different method for each of the subcommands that are then called here in the when statement, rather than doing all the logic for all subcommands here. Something that I was going to mention before, but forgot. This does mean rather than reusing the guild variable like I suggested in a previous comment, you may have to switch back to using event.guild!! each time like the others do as well.

tarna commented on src/main/kotlin/me/santio/minehututils/sticky/StickyManager.kt:
```diff
@@ -0,0 +1,136 @@
+    fun set(channelId: String, message: String) {
+        val sticky = stickyMessages[channelId]
+        if (sticky == null) {
+            start(channelId, message)
```
Rather than starting and immediately stopping when you want to set a message, it may be better to just run the same code that is in the sticky == null part of the start method, but active set to false, which is its default anyway.

tarna commented on src/main/kotlin/me/santio/minehututils/sticky/StickyManager.kt:
```diff
@@ -0,0 +1,136 @@
+            start(channelId, message)
+            stop(channelId)
+        } else {
+            stickyMessages[channelId]?.message = message
```
You should just do `sticky.message = message` here as you already have that variable set above and is known to not be null.

DDbrother9999 committed:
fix: various fixes

seperate when statement into functions

add forceRefresh

fix deleting message manually disabling the loop

null safety fixes

DDbrother9999 committed:
runcatching

DDbrother9999 committed:
fix: cubic return@forEach
